code/game/objects/items/rcd/RPD.dm

```diff
@@ -273,15 +298,114 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
 /obj/item/pipe_dispenser/attack_self(mob/user)
 	ui_interact(user)
 
-/obj/item/pipe_dispenser/pre_attack_secondary(obj/machinery/atmospherics/target, mob/user, params)
+/obj/item/pipe_dispenser/AltClick(mob/user)
+	. = ..()
+	if(!(upgrade_flags & RPD_UPGRADE_SPEEDPIPE))
+		return .
+	toggle_speedpipe_marking(user)
+
+/obj/item/pipe_dispenser/alt_click_secondary(mob/user)
+	. = ..()
+	speedpipe_build(user)
+
+/obj/item/pipe_dispenser/CtrlClick(mob/user)
+	. = ..()
+	clear_out_speedpipe(user)
+
+/obj/item/pipe_dispenser/proc/toggle_speedpipe_marking(mob/user)
+	speedpipe_marking = !speedpipe_marking
+	balloon_alert(user, "[speedpipe_marking ? "marking speedpipe locations" : "stopped marking"]")
+
+/obj/item/pipe_dispenser/proc/speedpipe_build(mob/user)
+	if(length(speedpipe_turfs))
+		speedpipe_building = !speedpipe_building
+		balloon_alert(user, "[speedpipe_building ? "speedpiping..." : "stopped speedpiping"]")
+		if(!speedpipe_building)
+			return
+		deltimer(timed_out_speedpipe_timer)
+		clear_orphaned_speedpipe_turfs()
+		for(var/turf/speedpipe_turf as anything in speedpipe_turfs)
+			if(!speedpipe_building)
```

Comment:
find a better way to do this. I don't like a for loop having a bunch of do afters AND checking an external var for state